src/tree.js

```diff
@@ -0,0 +1,44 @@
+import _ from 'lodash/fp'
+import {push} from './array'
+import {findIndexed} from './conversion'
+
+export let isTraversable = x => _.isArray(x) || _.isPlainObject(x)
+export let traverse = x => isTraversable(x) && !_.isEmpty(x) && x
+
+export let walk = (next = traverse) => (
+  pre,
+  post = _.noop,
+  parents = [],
+  parentIndexes = []
+) => (tree, index) =>
+  pre(tree, index, parents, parentIndexes) ||
+  findIndexed(
+    walk(next)(pre, post, [tree, ...parents], [index, ...parentIndexes]),
+    next(tree, index, parents, parentIndexes) || []
+  ) ||
+  post(tree, index, parents, parentIndexes)
```

Comment:
`pre`, `next` and `post` receive the same arguments. I wonder if it'd be better to make them an array and pass them with `...` 🤔 probably less performant, but would it look any better? just a thought.

Reply:
I think for now we can leave this as is - doing this would make it less performant and also add a line of code :)